### PR TITLE
feat(webhooks): send end user on connection creation

### PR DESCRIPTION
**Pull request by bodinsamuel**
## 🤓☝️ Changes

Fixes https://linear.app/nango/issue/NAN-1880/send-back-the-end-user-profile-in-the-webhook

- Send `endUser` in webhook on connection creation
When we use a session token, we pass metadata, but when we receive a webhook, we get a random connectionId that is impossible to link to an actual user. Sending back the userId (or orgId) will allow customers to link this connectionId to an actual user without needing to do any additional work.

NB: we don't send it in any other webhook since the connectionId should have been linked already

## 🧪 How to tests?

- Go to dashboard
- Modify webhook url to your favorite webhook dumper
- Allow webhook to be sent on connection creation
- Create a new connection with the connect UI

Screenshot 2024-11-26 at 17 32 59

### CODE REVIEW

File: packages/database/lib/getDbConfig.ts
Comment (bodinsamuel) [inline packages/database/lib/getDbConfig.ts:1]: Probably a leftover

File: packages/server/lib/utils/express.ts [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
+//           authType: 'publicKey';
+//           account: DBTeam;
+//           environment: DBEnvironment;
+//       }
+//     | {
+//           authType: 'basic' | 'session' | 'secretKey' | 'adminKey' | 'none';
+//           account?: DBTeam;
+//           environment?: DBEnvironment;
+//           user: Pick<DBUser, 'id' | 'email' | 'name'>;
+//       };
```
Comment (bodinsamuel) [inline packages/server/lib/utils/express.ts:22]: Tried to fix this type but opened a trap door (unfortunately I have wrote this type so I can only blame myself)
Will fix later

File: packages/server/lib/webhook/github-app-oauth-webhook-routing.ts [hunk 2]
```diff
@@ -98,7 +98,8 @@ async function handleCreateWebhook(integration: ProviderConfig, body: any, logCo
                     environment,
                     account,
                     auth_mode: 'APP',
-                    operation: res.operation
+                    operation: res.operation,
+                    endUser: undefined // TODO fix this
```
Comment (bodinsamuel) [inline packages/server/lib/webhook/github-app-oauth-webhook-routing.ts:102]: Github App webhook routing is weird, but it's an edge case

File: packages/shared/lib/models/Webhook.ts
Comment (bodinsamuel) [inline packages/shared/lib/models/Webhook.ts:1]: left over

**Review by nalanj — APPROVED** [review]
Looks good!

File: packages/server/lib/controllers/apiAuth.controller.ts [hunk 3]
```diff
@@ -156,7 +157,8 @@ class ApiAuthController {
                     environment,
                     account,
                     auth_mode: 'API_KEY',
-                    operation: updatedConnection.operation
+                    operation: updatedConnection.operation,
+                    endUser: isConnectSession ? res.locals['endUser'] : undefined
```
Comment (TBonnin) [inline packages/server/lib/controllers/apiAuth.controller.ts:161]: why not directly `res.locals['endUser']`? is it not `undefined` if not `isConnectSession`?

File: packages/server/lib/controllers/apiAuth.controller.ts [hunk 4]
```diff
@@ -28,10 +28,11 @@ import type { MessageRowInsert } from '@nangohq/types';
 
 class ApiAuthController {
     async apiKey(req: Request, res: Response<any, Required<RequestLocals>>, next: NextFunction) {
-        const { account, environment, authType } = res.locals;
+        const { account, environment } = res.locals;
         const { providerConfigKey } = req.params;
         const receivedConnectionId = req.query['connection_id'] as string | undefined;
         const connectionConfig = req.query['params'] != null ? getConnectionConfig(req.query['params']) : {};
+        const isConnectSession = res.locals['authType'] === 'connectSession';
```
Comment (TBonnin) [inline packages/server/lib/controllers/apiAuth.controller.ts:35]: `isConnectSession` could almost be an attribute of `RequestLocals`

File: packages/server/lib/utils/express.ts (hunk 1 above)
Comment (TBonnin) [reply packages/server/lib/utils/express.ts:22]: not sure I follow but I trust you

**Review by TBonnin — APPROVED** [review]

File: packages/server/lib/controllers/apiAuth.controller.ts (hunk 3 above)
Comment (bodinsamuel) [reply packages/server/lib/controllers/apiAuth.controller.ts:161]: true, I think I like the explicitness here and if (big if) I manage to fix the locals types `endUser` will be never in most case.

File: packages/server/lib/controllers/apiAuth.controller.ts (hunk 4 above)
Comment (bodinsamuel) [reply packages/server/lib/controllers/apiAuth.controller.ts:35]: maybe indeed

File: packages/server/lib/utils/express.ts (hunk 1 above)
Comment (bodinsamuel) [reply packages/server/lib/utils/express.ts:22]: This type is too loose meaning we can have connectSession with auth mode basic, which could lead to mistakes. Ideally, we have something stricter but when I wrote it I didn't anticipate this issue and now it seems some middleware does not enjoy the strictness when I tried to change it

**Pull request merged**